Dealing with long time length readings

kirtmd opened the pull request:
# Description

Dealing with long time length readings

Fixes #665, Fixes #806.

Changed the way that meter data is returned, previously OED would use daily and hourly readings before going to raw readings which lead to showing readings that didn't exist. With this code, OED starts by checking raw readings, then hourly and daily last to make sure that these issues don't persist. We added a reading frequency column to the meters table to help us do this. The function loops through each meter that is input because we need to know the frequency of each meter specifically to return it's proper reading interval.

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [ ] I have removed text in ( ) from the issue request

## Limitations

Issue #665 is still not completely finished, this covers the meter section of the issue but the group part is still TODO, you can read more about it here: https://github.com/OpenEnergyDashboard/DevDocs/blob/main/longReadings.md

thuloj committed:
Update create_reading_views.sql

thuloj committed:
Update create_reading_views.sql

kirtmd committed:
Added temporary frequency variable and implemented it. Also changed conditional statements to use the epoch of the interval to make it divisible

kirtmd committed:
Modified if statements and added a notice for testing

kirtmd committed:
Completed the function and cleaned up the code

kirtmd committed:
Created add_meter_reading_frequency.sql to add the frequency variable to meters that don't have it during migration.
Adjusted index.js accordingly and also deleted a few raise notice's from the create_reading_views.sql file.

kirtmd committed:
Added newline to alter table file and fixed a spacing error in create_reading_views.sql

huss commented on src/server/migrations/0.8.0-1.0.0/sql/meter/add_meter_reading_frequency.sql:
```diff
@@ -0,0 +1,7 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+-- Add unit_id and default_graphic_unit to the meters table.
```
I think this comment was copied from another file and needs to be updated.

huss commented on src/server/migrations/0.8.0-1.0.0/sql/meter/add_meter_reading_frequency.sql:
```diff
@@ -0,0 +1,7 @@
+
+-- Add unit_id and default_graphic_unit to the meters table.
+ALTER TABLE meters
+    ADD COLUMN IF NOT EXISTS meter_reading_frequency INTERVAL;
```
There is a default value in the meters creation SQL that should also be here.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -209,84 +210,87 @@ CREATE OR REPLACE FUNCTION meter_line_readings_unit (
 	unit_column INTEGER;
-BEGIN
+	frequency INTERVAL;
+	add INTEGER := 1;
```
Since add is the current meter id, should the name reflect that. meter_id is okay but seems close to meter_ids. Open to ideas but what about current_meter_id?

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -209,84 +210,87 @@ CREATE OR REPLACE FUNCTION meter_line_readings_unit (
-	 END IF;
+	-- Start with the raw, then hourly and then daily if others will not work
+
+	while add <= cardinality(meter_ids) loop
```
While you nicely put comments below, I think some more, esp. on what the next few lines are doing would help future developers.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -209,84 +210,87 @@ CREATE OR REPLACE FUNCTION meter_line_readings_unit (
+				r.start_timestamp,
+				r.end_timestamp
+				FROM (((readings r
+				--INNER JOIN unnest(current_id) meters(id) ON r.meter_id = meters.id)
```
Should this commented out line be removed?

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -209,84 +210,87 @@ CREATE OR REPLACE FUNCTION meter_line_readings_unit (
+
+	while add <= cardinality(meter_ids) loop
+		select meter_reading_frequency into frequency from meters where id = meter_ids[add];
+		IF (((EXTRACT (DAY FROM requested_interval) * 1440) / (EXTRACT (HOUR FROM frequency) * 60 + EXTRACT (MINUTE FROM frequency))) <= 1440) 
```
I cannot convince myself that the logic precisely does what is wanted but is close. It seems to assume that the requesteed_interval is an even number of days and I don't think that is the case. I think this since I thought EXTRTACT DAY only returns whole days. I also think we could have an issue with frequency if someone entered it in the general Postgres syntax of '4 days' rather than '96:00:00' since it does not deal with days. A quick test found this did not work with the current code.

To do this in a general way is probably desirable given that there are variety of ways to enter these values into Postgres. I'm hoping there is a way to do this but I have not investigated.

huss reviewed (changes requested):
Thanks to @kirtmd for putting in their first PR. I have left some comments to think about and address.

I did modest testing and found this generally works as desired. More substantial testing will be needed after it is finalized and before merging.

I also noticed it has conflicts with development over other changes that need to be resolved. I can help with that if you want.

Also see a comment on this PR (about to do) about other changes needed in OED before this can be deployed.

huss commented:
Working with this PR and thinking some more lead me to realize that the following changes are needed in OED before this can be deployed:

- Changing the way meters gets data impact groups (that use the new code for each meter in a group) so the group code needs to be updated. It needs to be the case that all meters are getting readings at the same frequency so they can be summed. We have to decide which of the possible meter frequencies to use but the one with the fewest points would be the conservative choice. It also needs to exclude raw. All of this will likely mean changes/new parameters for the line meter function.
- The meter pages need to be changed to allow interaction with the meter frequency of reading.
- The CSV meter import needs to work with the meter frequency of reading (might work okay but need to check).
- Many automated test cases that looks at line readings returned will need to be updated to reflect the new values returned.
- The test data and website data need to be updated to include the correct frequency of reading for each meter. This will also require modifying the meter input code to handle this.

Given that OED v1.0 should happen soon and the changes should be done before this PR is included in an OED release, this PR is on hold for merging to make sure either it happens after v1.0 or all changes are in place before merging.

Please note this does not reflect on the work done for the PR but issues outside of this work that must be accomplished. This PR can still progress with changes and be ready to merge once all the other work is done.

huss committed:
bug and comments

The test for daily was changed and comments were added.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/kirtmd/828

huss committed:
resolve migration comments on meter

huss committed:
resolve most comments on line reading SQL

huss committed:
fix syntax mistakes

huss committed:
update SQL, not full check yet

- This changes how the checks are done so they should work with
any Postgres interval.
- Changed so now calculates the range of readings for
each meter. This means you get the optimal for each meter
but it may be slower. Need to test for speed.
- Some cleanup of variable names, etc.

huss committed:
uppercase SQL keywords

huss committed:
rename table column

Others do not have meter so remove here. Also some comments.

huss committed:
update model, fix tests

- Add reading frequency to model and other SQL functions.
- Fix 23 min. test to use correct reading frequency.
- Fix 15 min test for hourly boundary since allow 1 more point

huss committed:
control line reading accuracy

Add enum and parameter to allow called to fix the accuracy of points
returned for line readings. First step toward making groups work.

huss committed:
integrate groups

Groups now work where do all meters as either hourly or daily.

huss committed:
graphing works

- Pass arguments to line SQL functions to control # of points.
- Readings requests use max points and 'auto' for new SQL.
- New function to set max points on server. Moved to util/ since seems
better place and not in sql/. Remove old file/function.
- Fix group reading test because now gets hourly not raw.
- Fix mistake on new enum location.
- Fix bug where line state was not set on certain zooms.
- Resolve TODO that was wrong.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/kirtmd/828

huss committed:
admin default meter frequency added

The trickier part was getting the Postgres interval to display in the
desired way.

huss committed:
sync line reading migrations with changes

huss committed:
remove reading rate environment

- Remove the old OED_SITE_READING_RATE and TEST one since we allow for
site default and meter level setting of this value.
- Meter constructor sets the default readingFrequency if not provided.
This is not perfect but there is a comment. Mostly needed for testing.
- Meter table changed to not allow null reading_frequency since needs to
be set for other code to work.
- Update some tests, as needed, to set the reading frequency.
One test no longer works as desired so commented out until correct
test is worked out.

huss committed:
state holds user friendly defaultMeterReadingFrequency

The Redux state now holds defaultMeterReadingFrequency in the user
friendly format rather than the DB format. This removes the hack to
format the value on edit. The routes for get/set preferences does this
so even on edit the state gets what the DB thought the new value was
just in case it differs from what the user set. It also means that the
value shown on page is the standard one and not the format the user
input the value in after save.

huss committed:
mostly working

- meter create/edit now deals with reading frequency.
- Now DB returns updated items on meters so can update
state without reloading meter data.
- The redux state holds the user friendly reading frequency
since only used for display.
- Fix bug I introduced where put no unit translated instead of
unit allowed for meter unit.
- formatting, use ===/!==.

huss committed:
update meter creation

new Meter() calls now set reading frequency.
A few other small changes including adding area unit and reading
frequency to meter equivalency test.

huss committed:
update how load meter data

Instead of a long list of items in an array, now uses key:value pairs.

huss committed:
add freq & area to test data for meters

huss commented:
I have worked on the necessary changes. The important ones to note are:

- The comments on code in the PR are all addressed.
- Each meter correctly calculates the number of points for its readings in DB call. In the past it was for all meters.
- The DB now uses passed parameters to determine when to switch between raw, hourly and daily. It uses 1440 points for now.
- Groups now work with the new code where all meters return the same frequency so they can be added and raw is excluded.
- The preferences page allows setting of a default meter frequency reading. The old environment variable is removed.
- The meter page allows setting of the reading frequency.
- Changes integrated in all graphing pages.
- All meter creation and tests were updated for the changes in this PR.
- Changed so meter create/update returns the id and reading frequency. This is routed back to the client to set state. This means a reload of all meter state is no longer needed.
- A bug was fixed that messed up the meter unit choice on the meters page.
- A bug was fixed where line state was sometimes unset and caused an error.
- The test data for developers and web were updated for the new code. This includes changing the meter loading of test data to be by key:value pairs. The data now sets the reading frequency & area but not area for web since need to figure out what to do. Note groups have not been converted so their area is not set.

Note that a default of 15 minutes is used for reading frequency. A lower value was used since it will return fewer points and be faster if the admin does not set.

With these changes I think this PR is ready to consider for merging into OED 1.0.

spearec reviewed (approved):
I've looked over the code and everything looks good. I don't have a perfect understanding of the SQL logic used, but the comments were enough to give me a passable understanding of what is going on.

I wasn't able to test the changes much, but simply graphing the test4DaySin meter in OED is enough to immediately see that the changes seem to fix the issue at least in this case.

The only thing I noted that might be an issue is that reading_line_accuracy is not defined in a migration, but I'm not sure if this is needed depending on when the enum is created in OED.

huss committed:
add migration for reading_line_accuracy

Addresses PR review comment.

huss commented:
> The only thing I noted that might be an issue is that reading_line_accuracy is not defined in a migration, but I'm not sure if this is needed depending on when the enum is created in OED.

Thanks for catching that. A coming commit fixes it.

huss committed:
Merge remote-tracking branch 'upstream/development' into pr/kirtmd/828-1

huss commented:
Thanks to @spearec for jumping in and doing this review. I am taking it as approval. I should have addressed the one comment. It should be ready to merge.
